### PR TITLE
add proxy options for RTM connection

### PR DESCRIPTION
**Pull request by dispensable**
This PR add a proxy option for connection to slack with RTM.

ref:
  - slack doc: https://slack.dev/node-slack-sdk/rtm-api#proxy-requests-with-a-custom-agent
  - https-proxy-agent: https://www.npmjs.com/package/https-proxy-agent  

usage:
```yaml
proxy:
    enable: enable
    host: "example-proxy.host"
    port: 8118
    protocol: "http"
    headers: {
        "foo": "bar"
    }
```

I am not a professional TypeScript programmer, so any advices will be appreciated !

### CODE REVIEW

**Comment by Half-Shot** [comment]
Hi, please ensure you've read through https://github.com/matrix-org/matrix-appservice-slack/blob/master/CONTRIBUTING.md and in particular looked at the steps in https://github.com/matrix-org/matrix-appservice-slack/blob/master/CONTRIBUTING.md#making-pull-requests.

File: config/config.sample.yaml [hunk 1]
```diff
@@ -74,3 +74,12 @@ matrix_admin_room: "!aBcDeF:matrix.org"
 # Enable metrics reporting on http://0.0.0.0:bridgePort/metrics which can be scraped by prometheus
 # Optional
 enable_metrics: true
+
+# proxy for slack RTM connection
+proxy:
```
Comment (Half-Shot) [inline config/config.sample.yaml]: This would be better off as a parent of the `rtm` object, imo.

File: src/SlackRTMHandler.ts [hunk 2]
```diff
@@ -116,6 +118,7 @@ export class SlackRTMHandler extends SlackEventHandler {
         if (!botToken.startsWith("xoxb")) {
             throw Error("Bot token invalid, must start with xoxb");
         }
+
```
Comment (Half-Shot) [inline src/SlackRTMHandler.ts]: Extra newline?

File: src/SlackRTMHandler.ts [hunk 3]
```diff
@@ -149,7 +152,8 @@ export class SlackRTMHandler extends SlackEventHandler {
         const LOG_LEVELS = ["debug", "info", "warn", "error", "silent"];
         const connLog = Logging.get(`RTM-${logLabel.substr(0, LOG_TEAM_LEN)}`);
         const logLevel = LOG_LEVELS.indexOf(this.main.config.rtm!.log_level || "silent");
-        const rtm = new RTMClient(token, {
+
+        let rtmOpts = {
```
Comment (Half-Shot) [inline src/SlackRTMHandler.ts]: `const` please

File: src/SlackRTMHandler.ts [hunk 4]
```diff
@@ -160,7 +164,21 @@ export class SlackRTMHandler extends SlackEventHandler {
                 info: logLevel <= 2 ? connLog.info.bind(connLog) : () => {},
                 error: logLevel <= 3 ? connLog.error.bind(connLog) : () => {},
             } as Logger,
-        });
+        }
+
+        if (this.main.config.proxy.enable) {
+            const opts: HttpsProxyAgentOptions = {
+                "host": this.main.config.proxy.host,
```
Comment (Half-Shot) [inline src/SlackRTMHandler.ts]: All of these can be unquoted :)

File: src/SlackRTMHandler.ts [hunk 5]
```diff
@@ -160,7 +164,21 @@ export class SlackRTMHandler extends SlackEventHandler {
                 info: logLevel <= 2 ? connLog.info.bind(connLog) : () => {},
                 error: logLevel <= 3 ? connLog.error.bind(connLog) : () => {},
             } as Logger,
-        });
+        }
+
+        if (this.main.config.proxy.enable) {
+            const opts: HttpsProxyAgentOptions = {
```
Comment (Half-Shot) [inline src/SlackRTMHandler.ts]:
```suggestion
            const proxyOpts: HttpsProxyAgentOptions = {
```

File: src/SlackRTMHandler.ts [hunk 6]
```diff
@@ -160,7 +164,21 @@ export class SlackRTMHandler extends SlackEventHandler {
...
+                "host": this.main.config.proxy.host,
+                "port": this.main.config.proxy.port,
+                "protocol": this.main.config.proxy.protocol,
+                "headers": this.main.config.proxy.headers
+            }
+
+            const agent = new HttpsProxyAgent(opts);
+            rtmOpts["agent"] = agent;
```
Comment (Half-Shot) [inline src/SlackRTMHandler.ts]:
```suggestion
            rtmOpts.agent = new HttpsProxyAgent(opts);;
```

File: src/SlackRTMHandler.ts [hunk 7]
```diff
@@ -160,7 +164,21 @@ export class SlackRTMHandler extends SlackEventHandler {
...
+
+        if (this.main.config.proxy.enable) {
+            const opts: HttpsProxyAgentOptions = {
+                "host": this.main.config.proxy.host,
+                "port": this.main.config.proxy.port,
+                "protocol": this.main.config.proxy.protocol,
+                "headers": this.main.config.proxy.headers
+            }
+
+            const agent = new HttpsProxyAgent(opts);
```
Comment (Half-Shot) [inline src/SlackRTMHandler.ts]:
```suggestion
```

**Comment by auscompgeek** [comment]
Why should the proxy be separate for RTM and API requests?

**Comment by dispensable** [comment]
I will resubmit a new pr after studying typescript seriously, thank you all for taking time to review these codes <3

**Pull request closed without merging**